### PR TITLE
Implements the `Eq` trait for `Option`

### LINKED ISSUE #3542: Implement the `Eq` trait for `Option`
The `Identity` implements the `Eq` trait: 
https://github.com/FuelLabs/sway/blob/716b34f7a22ff2dbbcc2505debcdb5892aaedf7c/sway-lib-std/src/identity.sw#L13-L21

I would like the same thing to exist for `Option` so I can compare Options.

### PR DESCRIPTION
**Pull request by brandonsurh**
## Description
Closes #3542
Implements the `Eq` trait for `Option`. Also adds tests for the implementation.

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Comment by brandonsurh** [comment]
Will need help with adding assignees, adding labels, and requesting reviewers due to permissions.

**Comment by brandonsurh** [comment]
Not sure why the `cargo-run-e2e-test` check is failing. However, it doesn't not seem related to the contents of this PR.

**Comment by bitzoic** [comment]
> Not sure why the `cargo-run-e2e-test` check is failing. I don't think it's related to these changes, but I'm not certain.

Looks like this is a breaking change. You will need to replace the contract address in the e2e test script with `0xe2f370fb01c8c36a521093494a023ce2d7232398e5e88722164c42ec0303f381`.

Screenshot 2023-11-27 at 10 02 01 AM

**Review by Braqzen — CHANGES_REQUESTED** [review]
I don't believe that we need to check the permutations anymore and all of these negations can probably be cleaned up too.

Besides that, I need to block this until all primitives are tested. An integer and bool are insufficient here despite the trait constraint.

Once all types are tested this will be good to merge

**Comment by Braqzen** [comment]
> > Not sure why the `cargo-run-e2e-test` check is failing. I don't think it's related to these changes, but I'm not certain.
> 
> Looks like this is a breaking change. You will need to replace the contract address in the e2e test script with `0xe2f370fb01c8c36a521093494a023ce2d7232398e5e88722164c42ec0303f381`.
> 
> Screenshot 2023-11-27 at 10 02 01 AM

I don't believe this to be a breaking change. It extends the functionality but does not change the API.
Something somewhere else breaks, that's not a breaking change. That's a side effect of how we wrote the test suite. Once the API is changed for the user (ex. swap the order of parameters in a function) then that qualifies as a breaking change.

**Comment by bitzoic** [comment]
> > > Not sure why the `cargo-run-e2e-test` check is failing. I don't think it's related to these changes, but I'm not certain.
> > 
> > 
> > Looks like this is a breaking change. You will need to replace the contract address in the e2e test script with `0xe2f370fb01c8c36a521093494a023ce2d7232398e5e88722164c42ec0303f381`.
> > Screenshot 2023-11-27 at 10 02 01 AM
> 
> I don't believe this to be a breaking change. It extends the functionality but does not change the API. Something somewhere else breaks, that's not a breaking change. That's a side effect of how we wrote the test suite. Once the API is changed for the user (ex. swap the order of parameters in a function) then that qualifies as a breaking change.

Any changes that can change an existing contract's bytecode and hash must be labeled as a breaking change.

**Comment by Braqzen** [comment]
> > > > Not sure why the `cargo-run-e2e-test` check is failing. I don't think it's related to these changes, but I'm not certain.
> > > 
> > > 
> > > Looks like this is a breaking change. You will need to replace the contract address in the e2e test script with `0xe2f370fb01c8c36a521093494a023ce2d7232398e5e88722164c42ec0303f381`.
> > > Screenshot 2023-11-27 at 10 02 01 AM
> > 
> > 
> > I don't believe this to be a breaking change. It extends the functionality but does not change the API. Something somewhere else breaks, that's not a breaking change. That's a side effect of how we wrote the test suite. Once the API is changed for the user (ex. swap the order of parameters in a function) then that qualifies as a breaking change.
> 
> Any changes that can change an existing contract's bytecode and hash must be labeled as a breaking change.

I'm unsure whether I agree with that. Can you tell me where that reasoning comes from and where this was brought up?

**Comment by brandonsurh** [comment]
> I don't believe that we need to check the permutations anymore and all of these negations can probably be cleaned up too.
> 
> Besides that, I need to block this until all primitives are tested. An integer and bool are insufficient here despite the trait constraint.
> 
> Once all types are tested this will be good to merge

Cleaned up the permutations and negations. Tests were rewritten and now include the following types that were listed as primitives in the Sway Book:

- u8
- u16
- u32
- u64
- u256
- str
- str[]
- bool
- b256

The only problem is with `Eq` trait when testing the `str[]` (String Array) type. The last commit gives the following error:
```
 error
             --> /home/brandon/sway/test/src/e2e_vm_tests/test_programs/should_pass/stdlib/option_eq/src/main.sw:119:24
              |
          117 |
          118 |     // Eq tests
          119 |     assert(str_option1 == str_option1);
              |                        ^^ Trait "Eq" is not implemented for type "str[4]".
          120 |     assert(str_option1 == str_option2);
              |
          ____
```

Not sure what is causing this, but will look into it. It's possible that I am just handling it incorrectly.

**Comment by brandonsurh** [comment]
Decided to skip over the fixed-length string tests and remove them. They are not within the constraints of the `Eq` trait. Tests should be good now.

**Review by bitzoic — DISMISSED** [review]

**Review by Braqzen — COMMENTED** [review]
Struct, enum, tuple and array tests?

**Comment by brandonsurh** [comment]
> Struct, enum, tuple and array tests?

Wasn't sure if these were within the scope for testing because they aren't primitive types. Though you're right, it's probably a good idea to test them anyways. Tests with them pass, however, implementing the `eq` method for them come with some warnings about them not being used for some reason. They are most definitely called because if you remove them then the tests break. Because of this, I added `expected_warnings = 5` to the `test.toml` file.

```
          warning
            --> /home/brandon/sway/test/src/e2e_vm_tests/test_programs/should_pass/stdlib/option_eq/src/main.sw:14:5
             |
          12 |
          13 |   impl core::ops::Eq for StringArray {
          14 |       fn eq(self, other: Self) -> bool {
             |  _____-
          15 | |         from_str_array(self) != from_str_array(other)
          16 | |     }
             | |_____- This method is never called.
          17 |   }
             |
          ____

          warning
             --> /home/brandon/sway/sway-lib-core/src/ops.sw:536:5
              |
          534 |
          535 |       /// ```
          536 |       fn neq(self, other: Self) -> bool {
              |  _____-
          537 | |         (self.eq(other)).not()
          538 | |     }
              | |_____- This method is never called.
          539 |   }
              |
          ____

          warning
            --> /home/brandon/sway/test/src/e2e_vm_tests/test_programs/should_pass/stdlib/option_eq/src/main.sw:21:5
             |
          19 |
          20 |   impl core::ops::Eq for Array {
          21 |       fn eq(self, other: Self) -> bool {
             |  _____-
          22 | |         self[0] == other[0] && self[1] == other[1]
          23 | |     }
             | |_____- This method is never called.
          24 |   }
             |
          ____

          warning
            --> /home/brandon/sway/test/src/e2e_vm_tests/test_programs/should_pass/stdlib/option_eq/src/main.sw:27:5
             |
          25 |
          26 |   impl core::ops::Eq for MyStruct {
          27 |       fn eq(self, other: Self) -> bool {
             |  _____-
          28 | |         self.x == other.x
          29 | |     }
             | |_____- This method is never called.
          30 |   }
             |
          ____

          warning
            --> /home/brandon/sway/test/src/e2e_vm_tests/test_programs/should_pass/stdlib/option_eq/src/main.sw:34:5
             |
          32 |
          33 |   impl core::ops::Eq for Tuple {
          34 |       fn eq(self, other: Self) -> bool {
             |  _____-
          35 | |         self.0 == other.0 && self.1 == other.1
          36 | |     }
             | |_____- This method is never called.
          37 |   }
             |
          ____

            Compiled script "option_eq" with 5 warnings.
```

**Review by Braqzen — APPROVED** [review]

File: test/src/e2e_vm_tests/test_programs/should_pass/stdlib/option_eq/test.toml [hunk 1]
```diff
@@ -1,3 +1,5 @@
 category = "run"
 expected_result = { action = "return", value = 1 }
 validate_abi = true
+
+expected_warnings = 5
```
[the remark this answers is not in the document]
Comment (brandonsurh) [reply test/src/e2e_vm_tests/test_programs/should_pass/stdlib/option_eq/test.toml:5]: That's a good evaluation. I think that would be the case too. The problem is that if there are warnings without this flag then the test fails. I don't think the testing suite accepts when there are warnings without them being specified in this manner.

**Review by IGI-111 — APPROVED** [review]

**Comment by brandonsurh** [comment]
The recent PR #5210 that changes the way `impl` trait functions are handled breaks this PR. Will address this.

**Pull request merged**